src/js/common/utils/accessible-menus.js

```diff
@@ -0,0 +1,349 @@
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // Close the menu
+  menuButton.removeAttribute('aria-expanded');
+  menu.setAttribute('hidden', 'hidden');
+}
+
+
+/**
+ * Opens a menu.
+ *
+ * @param {HTMLLIElement} menuLi  The <li> containing the menubutton and menu
+ * @param {bool} openSubMenu      Whether or not the first submenu should be opened. E.g. if we're opening the
+ *                                 second-level menu in a wide screen, we want the third-level opened as well
+ * @param {bool} stealFocus       Whether or not to focus on the first item in the opened menu
+ */
+function openMenu(menuLi, openSubMenu = false, stealFocus = true) {
+  // If we're not dealing with a menu structure, abort
+  const struct = getMenuStructure(menuLi);
+  if (!struct) {
+    return;
+  }
+
+  const { menuButton, menu } = struct;
+
+  // First, close all sibling menus
+  const openMenus = menuLi.parentElement.querySelectorAll('[aria-expanded=true]');
+  openMenus.forEach(m => closeMenu(m.parentElement));
+
+  // Open the menu
+  menuButton.setAttribute('aria-expanded', true);
+  menu.removeAttribute('hidden');
+
+  if (stealFocus) {
+    const menuRole = (menu.getAttribute('role') || '').toLowerCase();
+    const firstMenuItem = menuRole === 'menu' ?
+      firstVisibleChild(menu) :
+      firstVisibleChild(menu.querySelector('[role="menu"]'));
+    firstMenuItem.firstElementChild.focus();
+  }
+
+  // If we're wide-screen (for example), open the first submenu
+  if (openSubMenu) {
+    openMenu(menu.firstElementChild, false, false);
+  }
+}
+
+
+/**
+ * Closes all the menus in menuElement
+ *
+ * @param {HTMLElement} menuElement
+ * @return {function}   Call this to close all menus
+ */
+function closeAll(menuElement) {
+  // Get all the parent <li>s of open menus
+  const allMenus = Array.from(menuElement.querySelectorAll('[aria-expanded=true]'))
+    .map(e => findNearestLi(e));
+
+  // Add the top-level 
+  if (menuElement.getAttribute('aria-expanded') === true) {
+    allMenus.push(findNearestLi(menuElement));
+  }
+
+  allMenus.forEach(e => closeMenu(e));
+}
+
+
+/**
+ * Attaches event listeners to a menu or menu bar to make it keyboard navigable.
+ * 
+ * If the mobile buttons are provided, this will also handle the opening and closing
+ *  of the menu on small screens when there's just a "Menu" button.
+ *
+ * @param {HTMLElement} menuElement        A menubar or menu
+ * @param {HTMLElement} mobileOpenButton   Optional - The "Menu" button on mobile
+ * @param {HTMLElement} mobileCloseButton  Optional - The "Close" button on mobile
+ */
+export default function addMenuListeners(menuElement) {
+  const menuRole = menuElement.getAttribute('role');
+  if (!['menubar', 'menu'].includes(menuRole)) {
+    // If we don't have a menubar or menu, don't continue
+    return;
+  }
+
+  // For all the sub menus, add listeners for:
+  //  Up, down, left, right, escape
+  menuElement.addEventListener('keydown', (e) => {
```

Comment:
Could it be advantageous to attach separate listeners to menubar, menu, and menuitem or would that unnecessarily duplicate code?

Reply:
That's how I was originally going to do it, actually. As I was starting to implement it, however, I went this way because it seemed like a lot of code duplication. It might still make more sense to split them up, but...I'm not sure, honestly.

Reply:
Actually, now that I think about it, it might be best to not add it to the individual items to allow for items added to the menu dynamically after its creation. May not ever happen, but future proofing is rarely bad (if it's still performant and maintainable, of course).

Reply:
Right, event delegation would be used in that case, but the idea would be to separate the logic if desirable.